### PR TITLE
Add `resource` scheme for #11253

### PR DESCRIPTION
**Pull request by kijuky**
# Pull Request Checklist

* [x] Have you read [How to write the perfect pull request](https://github.com/blog/1943-how-to-write-the-perfect-pull-request)?
* [x] Have you read through the [contributor guidelines](https://www.playframework.com/contributing)?
* [x] Have you signed the [Lightbend CLA](https://www.lightbend.com/contribute/cla)?
* [x] Have you referenced any issues you're fixing using [commit message keywords](https://help.github.com/articles/closing-issues-using-keywords/)?
* [x] Have you added copyright headers to new files?
* [x] Have you checked that both Scala and Java APIs are updated?
* [x] Have you updated the documentation for both Scala and Java sections?
* [x] Have you added tests for any changed functionality?

# Helpful things

## Fixes

This pull request only supports the `resource` scheme. I will not add support for other graalvm features.

## Purpose

graalvm requires resources to be accessed with a scheme called `resource`. So Play needs to handle the `resource` scheme properly.

## Background Context

- The implementation has been copied over from the existing `bundleresource` scheme as it is functionally equivalent.
- Accessing the directory results in an invalid file download. This is because the `resource` scheme cannot tell if the file is a directory.

## References

Discussion for #11253

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
At least one commit author (kizuki-yasue@m3.com) is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

**Comment by kijuky** [comment]
Please help me. 😢 

Test case "play-sbt-plugin/shutdown-downing" is failing. However, I was unable to reproduce the failure when running at local or [at CI on a forked repository](https://github.com/kijuky/playframework/pull/1). Where is the cause...?

**Comment by mkurz** [comment]
@kijuky The `shutdown-downing` scripted test is flaky, it's not your fault 😉 I restarted it

**Comment by kijuky** [comment]
@mkurz 
Thank you very much! It helped me a lot.

**Comment by kijuky** [comment]
In the current implementation, accessing the directory results in invalid file download behavior. To avoid this, we need to add an Environment argument to the Assets constructor for directory detection.
https://github.com/playframework/playframework/compare/main...kijuky:add-resource-scheme:feature/add-resource-scheme-2

Would it be better to include the diff in this pull request?

File: core/play/src/main/scala/play/api/controllers/Assets.scala [hunk 1]
```diff
@@ -573,6 +573,7 @@ private class AssetInfo(
       case "file"                      => Some(httpDateFormat.format(Instant.ofEpochMilli(new File(url.toURI).lastModified)))
       case "jar"                       => getLastModified[JarURLConnection](c => c.getJarEntry.getTime)
       case "bundle" | "bundleresource" => getLastModified[URLConnection](c => c.getLastModified)
+      case "resource"                  => getLastModified[URLConnection](c => c.getLastModified)
```
Comment (gmethvin) [inline core/play/src/main/scala/play/api/controllers/Assets.scala:576]: would be cleaner to do `case "bundle" | "resource" | "bundleresource" => ...`

**Comment by gmethvin** [comment]
@kijuky did you want to add the other changes in this PR? They're related so I think it's fine. But if you prefer you can open a different PR.

**Comment by mkurz** [comment]
@kijuky Are you able to continue to work on this pull request? Like Greg said, just add the other changes as well, it's fine. Thanks!

**Comment by kijuky** [comment]
@gmethvin @mkurz Sorry for being late.

> did you want to add the other changes in this PR? They're related so I think it's fine. But if you prefer you can open a different PR.

Thank you for your advice. Added a commit that fixed the directory issue. This required including the `Environment` in the `Assets` class. So more fixes were needed.

File: core/play/src/main/scala/play/utils/Resources.scala [hunk 2]
```diff
@@ -20,32 +20,37 @@ object Resources {
...
-        }
-      }
-    case other => false
+  def isUrlConnectionADirectory(classLoader: ClassLoader, urlConnection: URLConnection): Boolean = {
+    (urlConnection, urlConnection.getURL.getProtocol) match {
+      case (file, "file") =>
+        new File(file.getURL.toURI).isDirectory
+      case (jar: JarURLConnection, _) =>
+        isJarURLConnectionDirectory(jar)
+      case (bundle, "bundle") =>
```
Comment (gmethvin) [inline core/play/src/main/scala/play/utils/Resources.scala:45]: For all of these other cases I think we can just call `isDirectory(urlConnection.getURL)`. That'd make the code a bit easier to understand.

**Comment by mkurz** [comment]
@kijuky Can you have a look at @gmethvin comments?

**Review by kijuky — COMMENTED** [review]
Very sorry... I didn't run "review changes".

File: core/play/src/main/scala/play/utils/Resources.scala [hunk 3]
```diff
@@ -20,32 +20,37 @@ object Resources {
     case "jar"                       => isZipResourceDirectory(url)
     case "zip"                       => isZipResourceDirectory(url)
     case "bundle" | "bundleresource" => isBundleResourceDirectory(classLoader, url)
+    case "resource"                  => isGraalVMResourceDirectory(classLoader, url)
```
[the remark this answers is not in the document]
Comment (gmethvin) [reply core/play/src/main/scala/play/utils/Resources.scala:23]: Thanks for the explanation; it makes more sense now that you've changed the names in the code and added a comment to clarify.

**Review by gmethvin — APPROVED** [review]

File: core/play/src/main/scala/play/api/controllers/Assets.scala [hunk 4]
```diff
@@ -821,7 +832,7 @@ class AssetsBuilder(errorHandler: HttpErrorHandler, meta: AssetsMetadata) extend
       case Some((assetInfo, acceptEncoding)) =>
         val connection = assetInfo.url(acceptEncoding).openConnection()
         // Make sure it's not a directory
-        if (Resources.isUrlConnectionADirectory(connection)) {
+        if (env != null && Resources.isUrlConnectionADirectory(env.classLoader, connection)) {
```
Comment (mkurz) [inline core/play/src/main/scala/play/api/controllers/Assets.scala:835]: For backwards compatibility, shouldn't we keep calling `Resources.isUrlConnectionADirectory` even when env is null?
```suggestion
        if (Resources.isUrlConnectionADirectory(if (env != null) env.classLoader else null, connection)) {
```
Makes sense IMHO.

**Review by mkurz — CHANGES_REQUESTED** [review]
@kijuky LGTM besides one comment. If we solve that that PR can be merged.

File: core/play/src/main/scala/play/api/controllers/Assets.scala (hunk 4 above)
Comment (kijuky) [reply core/play/src/main/scala/play/api/controllers/Assets.scala:835]: I see. Certainly, even if it already exists, there is a path in the calling hierarchy where classLoader is assumed to be null. I will reflect your suggestions.

**Review by mkurz — APPROVED** [review]
Thanks!

**Pull request merged**